gradle/scripts/lib/java-javadoc.gradle

```diff
@@ -316,126 +341,160 @@ allprojects {
+static def normalizeJavadocUrl(String javadocUrl) {
+    if (!javadocUrl.endsWith('/')) {
+        javadocUrl += '/'
+    }
+    return javadocUrl
+}
 
-                    conn.disconnect()
-                } catch (e) {
-                    tmpListFile.delete()
-                    logger.log(LogLevel.WARN, "Download failed: ${e}", e)
-                }
+@CacheableTask
+class OfflineLinkDownloadTask extends DefaultTask {
 
-                return success
-            }
+    @Input
+    Map<String, String> links = [:]
 
-            def addOfflineLink = { name, url ->
-                if (offlineJavadoc) {
-                    return
-                }
+    @Input
+    JavaVersion javaVersion = JavaVersion.current()
 
-                def javadocUrl = "${url}"
-                if (!javadocUrl.endsWith('/')) {
-                    javadocUrl += '/'
-                }
+    @Internal
+    File javadocCacheDir
 
-                def javadocUrlSha1 = MessageDigest.getInstance('SHA1').digest(javadocUrl.getBytes('UTF-8')).encodeHex()
-                def listFileDir = new File(javadocCacheDir, "${name}/${javadocUrlSha1}")
-                def packageListFile = new File(listFileDir, 'package-list')
-                def elementListFile = new File(listFileDir, 'element-list')
+    @OutputFile
+    File elementListFile
 
-                if (packageListFile.exists() && packageListFile.length() == 0) {
-                    packageListFile.delete()
-                }
-                if (elementListFile.exists() && elementListFile.length() == 0) {
-                    elementListFile.delete()
-                }
+    @OutputFile
+    File packageListFile
+
+    @Internal
+    def visitedUrls = new HashSet<>()
+
+    @Internal
+    Map<String, String> url2File = [:]
+
+    @Input
+    boolean offlineJavadoc
 
-                def success = packageListFile.exists() ||
-                              JavaVersion.current() >= JavaVersion.VERSION_1_10 && elementListFile.exists()
-                if (!success) {
-                    if (downloadListFile(packageListFile, new URL("${javadocUrl}package-list"))) {
-                        // Succeeded to download package-list.
+    @Inject
+    OfflineLinkDownloadTask(File javadocCacheDir, Map<String, String> links,
+                            boolean offlineJavadoc) {
+        this.links = links
+        this.javadocCacheDir = javadocCacheDir
+        packageListFile = new File(javadocCacheDir, 'package-list')
+        elementListFile = new File(javadocCacheDir, 'element-list')
+        this.offlineJavadoc = offlineJavadoc
+    }
+
+    @TaskAction
+    def run() {
+        if (offlineJavadoc) {
+            return
+        }
+
+        links.entrySet().forEach {
+            def (success, url, file) = addOfflineLink(it.key, it.value)
+            if (success) {
+                url2File[url] = file.toString()
+            }
+        }
+    }
+
+    def addOfflineLink(String name, String javadocUrl) {
```

Comment:
```suggestion
    private def addOfflineLink(String name, String javadocUrl) {
```

Reply:
Got this error message when trying to do so:

```
* What went wrong:
Execution failed for task ':downloadJavadocPackageList'.
> Could not find method addOfflineLink() for arguments [java11, https://docs.oracle.com/en/java/javase/11/docs/api/] on task ':downloadJavadocPackageList' of type DownloadJavadocPackageListTask.
```